src/subnet_rental_canister/src/canister.rs

```diff
@@ -74,77 +74,82 @@ fn start_timers() {
-        // Only try to lock if we haven't already locked 100% or more.
-        // Use multiplication result to account for rounding errors.
-        if locked_amount_icp >= Tokens::from_e8s(ten_percent_icp.e8s() * 10) {
-            println!("Rental request for {user} is already fully locked.");
-            continue;
-        }
+async fn lock_icp_for_rental_request(rental_request: RentalRequest, now_nanos: u64) {
+    let RentalRequest {
+        user,
+        initial_cost_icp,
+        locked_amount_icp,
+        locked_amount_cycles,
+        initial_proposal_id,
+        creation_time_nanos,
+        rental_condition_id,
+        last_locking_time_nanos,
+    } = rental_request;
 
-        let Ok(_guard_res) = CallerGuard::new(Principal::anonymous(), "rental_request") else {
-            println!("Busy processing another request. Skipping.");
-            continue;
-        };
+    // If the last locking time is less than 30 days ago, skip.
+    if (now_nanos - last_locking_time_nanos) / BILLION / 60 / 60 / 24 < 30 {
+        return;
+    }
 
-        // Convert ICP to cycles.
-        let locked_cycles =
-            match convert_icp_to_cycles(ten_percent_icp, Subaccount::from(user)).await {
-                Ok(locked_cycles) => locked_cycles,
-                Err(error) => {
-                    println!("Failed to convert ICP to cycles for rental request of user {user}.");
-                    persist_event(
-                        EventType::LockingFailure {
-                            user,
-                            reason: format!("{error:?}"),
-                        },
-                        Some(user),
-                    );
-                    continue;
-                }
-            };
-
-        println!("SRC gained {locked_cycles} cycles from the locked ICP.");
-        persist_event(
-            EventType::LockingSuccess {
-                user,
-                amount: ten_percent_icp,
-                cycles: locked_cycles,
-            },
-            Some(user),
-        );
+    let ten_percent_icp = Tokens::from_e8s(initial_cost_icp.e8s() / 10);
+    // Only try to lock if we haven't already locked 100% or more.
+    // Use multiplication result to account for rounding errors.
+    if locked_amount_icp >= Tokens::from_e8s(ten_percent_icp.e8s() * 10) {
+        println!("Rental request for {user} is already fully locked.");
+        return;
+    }
 
-        let locked_amount_icp = locked_amount_icp + ten_percent_icp;
-        let locked_amount_cycles = locked_amount_cycles + locked_cycles;
-        let new_rental_request = RentalRequest {
+    let Ok(_guard_res) = CallerGuard::new(Principal::anonymous(), "rental_request") else {
+        println!("Busy processing another request. Trying again now.");
+        // If we fail to acquire the lock, try again.
+        ic_cdk::futures::spawn(lock_icp_for_rental_request(rental_request, now_nanos));
```

Comment:
We should spawn in a global timer (in a separate call context executed later), not this way (otherwise there's no way for the lock to be released).